### ISSUE TITLE
Performance issue using Preact and 1.0.0-rc.3

### ISSUE THREAD

**Issue by rjyo** [chakra-ui/chakra-ui#2012]
# Bug report

## Describe the bug

I met this problem when using next.js with preact and 1.0.0-rc.(1|2|3). Everything works in dev mode but when the app is built and run in production mode, the page takes a very long time to load.

## To reproduce

I've build an demo for this problem. 

1. `npm i`
2. `npm run build`
3. `npm start`
4. Open `http://localhost:3000`

## Minimal reproduction

https://github.com/rjyo/demo-preact-chakraui

## Expected behavior

N/A

## Screenshots

A simple page with 6 text `Input` took several seconds before page loaded. 

![image](https://user-images.githubusercontent.com/352545/92927438-9027c600-f478-11ea-987a-f9776750e49e.png)

## System information

- OS: macOS Catalina 10.15.6
- Browser (if applies): Chrome
- Version of @chakra-ui/core: 1.0.0-rc.3
- Version of Node.js: v12.18.0

## Additional context

I know `Preact` many not be officially supported. Hope you guys have some time to look into this.

### How I found this

Comparing 2 simple "Terms of Usage" pages (mostly displaying text), one build with Chakra-UI 0.8 + Preact, another with Chakra-UI rc.3 + React, the score of Google PageSpeed Insights went down quite a lot. More specifically, the `Total Blocking Time` almost doubled. I thought it could be React's performance problem but found Chakra-UI rc.3 + Preact doesn't even work anymore.

**Comment by ljosberinn** [comment]
I've actually tried to get it running but the Preact example seems to be incompatible with Next 9.5 currently. It doesn't finish loading the index page for me. I'm fairly confident its either an issue with Preact + Next 9.5+ in general or a dependency of Chakra having issues with Preact.

**Comment by rjyo** [comment]
Hi @ljosberinn, thanks for looking into this issue.

I'm using next 9.5+ too and I think Next 9.5+ w/ Preact many not causing this problem.

I just pushed an update to the demo repo, with the index page changed to a normal speed page (with `Heading`, `Button`, `Link` and even `Toast`) and a link to the slow page (the page with 6 `Input`s).

I know it may not be just the `Input`'s problem. Just hope this simple demo can help tracing down the real problem.

**Comment by ljosberinn** [comment]
Thank you I will take a look tomorrow and do some profiling. I'd love to see Chakra being compatible with Preact as it's such an easy optimization to remove 30kb JS ☀️

**Comment by ljosberinn** [comment]
I can confirm it is indeed very slow. Now the interesting part:

- importing this into `_app.tsx` for debugging reasons, which effectively just enables preact devtools in prod fixes it.

```ts
if (typeof window !== 'undefined') {
  require('preact/debug')
}
```

Obviously that can't be the solution, so I also added `@zeit/next-source-maps` to the `next.config.js`. Using chromes Performance tab, I could track down the issue: `useStyleConfig` using `lodash.merge`.

![image](https://user-images.githubusercontent.com/29307652/92992159-285d9380-f4e9-11ea-96a5-d8e52e51a10e.png)

In your case, its twice as bad because youre passing a theme thats the default theme, which we then merge... with the default theme.
I _think_ we can omit that merge process just by comparing those.

Merging the theme with itself isn't horribly unperformant actually, but as you can see here, each `useStyleConfig('Form', ...)` takes ages. Really weird behaviour considering this only happens with Preact.

![image](https://user-images.githubusercontent.com/29307652/92992841-116d7000-f4ee-11ea-8450-019bc8c5a239.png)

Digging some more and will update.

**Comment by ljosberinn** [comment]
I copied the same code without omitting children into my project and it's not happening in React.

![image](https://user-images.githubusercontent.com/29307652/92993749-775cf600-f4f4-11ea-85fc-627e30ca3450.png)

**Comment by rjyo** [comment]
Wow! Congratulations! Can't wait to test the benchmark it again with the new PR!

Thanks so much for the quick and neat response!

---

> Edit: sadly cant compare outerTheme === theme in theme provider. but you shouldnt pass the default theme anyways, so that would be user error anyways. and 4ms isnt too bad.

As to theme merging, from the following code in the Getting Started section, does it means that even we merge the theme before passing it to `ChakraProvider`, there will be still another merge inside `ChakraProvider`?

```typescript
// 2. Extend the theme to include custom colors, fonts, etc.
const customTheme = merge(theme, {
  colors: {
    brand: {
      900: "#1a365d",
      800: "#153e75",
      700: "#2a69ac",
    },
  },
})

// 3. Pass the `theme` prop to the `ChakraProvider`
function App() {
  return (
    <ChakraProvider resetCSS theme={customTheme}>
      <App />
    </ChakraProvider>
  )
}
```

**Comment by ljosberinn** [comment]
Those docs are outdated, actually. They are updated in a different branch, `merge` was renamed to `extendTheme` and you don't need to pass the default theme yourself anymore.

If you don't customize the theme, you can omit it entirely, as it falls back to the default theme anyways: https://github.com/chakra-ui/chakra-ui/blob/1457992cfab149192f92ab7c4d687af97d6ffd43/packages/core/src/chakra-provider.tsx#L30

I'm probably wrong about comparing `outerTheme` with `theme` here https://github.com/chakra-ui/chakra-ui/blob/1457992cfab149192f92ab7c4d687af97d6ffd43/packages/system/src/providers.tsx#L17 as this seems rather to be for cases where you have multiple `ThemeProvider`. Initially, it will always be `merge({}, {}, defaultTheme)`.

And even if I'm not, `lodash.merge` should be clever enough to do a referential equality check before acting...

**Comment by ljosberinn** [comment]
On a sidenote - this probably won't fix all issues there are. I'm trying to move my project to preact and don't really get far. 

Weirdly enough, when I use this component:

```js
function FadedText({ children }: WithChildren) {
  const color = useColorModeValue('blackAlpha.700', 'whiteAlpha.700');

  return (
    <Text color={color} fontSize="xl" mt="6" px={2}>
      {children}
    </Text>
  );
}
```

RAM goes up to over 4GB and eventually crashes.

Using the component _inline_, works however.

**Comment by ljosberinn** [comment]
https://github.com/emotion-js/emotion/pull/896 that will probably explain it 😅 preact is not supported by emotion, should've searched earlier, oh well. At least we found that other issue because of this rabbit hole.

**Comment by ljosberinn** [comment]
Didn't mean to close it, the PR will do that, sorry.

**Comment by rjyo** [comment]
Hope that issue alone can improve my `TBT` score.

And I guess now the official answer to "Can I use Preact with Chakra?" is quite clear. ;)

**Comment by ljosberinn** [comment]
@rjyo I can't tell what exactly led to it but after testing again, it works now! Please try again in your repo after removing `node_modules` and lockfile and report back if you can :)

**Comment by rjyo** [comment]
@ljosberinn Thanks! I tried and it does work!

Beside the js bundle's weight lost of 32KB (around 30% in my case), I also got some performance boost when switching to Preact. Not so dramatical, but I think it worth the efforts.

<img width="400" alt="before" src="https://user-images.githubusercontent.com/352545/96196941-9f27f980-0f8b-11eb-8212-724eda5c5139.png"> <img width="400" alt="after" src="https://user-images.githubusercontent.com/352545/96196951-a222ea00-0f8b-11eb-9cb6-3b2d87619f65.png">

Also I've tried to remove `@prefresh/next` as you've mentioned. In dev mode, without it the pages won't be correctly reloaded. So I guess it's still a needed dependency when you also use Preact in dev mode.
